Add lexer for the Clean language

camilstaps opened the pull request:
Some notes:

- [Clean](http://clean.cs.ru.nl) is a functional programming language similar to Haskell. Because Clean uses different comment markers, Haskell lexers will immediately fail on almost any Clean code.
- Some regexes in here were stolen from https://gitlab.science.ru.nl/cloogle/clean-highlighter/blob/master/clean.js which I maintain. It attempts to recognise many more fine distinctions but is because of that unfortunately more buggy. In here I tried to be more minimal.
- The `code { ... }` construct allows one to use inline ABC code (comparable to inline assembly code in C).

Any suggestions would be welcome! Thanks for taking the time to review :bowing_man:

pyrmont commented:
@camilstaps Thanks for the PR! I'll try to look at this over the next couple of days :)

camilstaps commented:
That's great, @pyrmont! You have done an amazing job with catching up with the backlog so far, as far as I can see :+1:

pyrmont commented on spec/visual/samples/clean:
```diff
@@ -0,0 +1,224 @@
+implementation module Data.GenCompress
+
+// This sample file was taken from Clean Platform:
```
Real world samples can be helpful but a fair amount of the code here seems to be duplicative (from a testing point of view). Would it be possible to reduce some of that so that each line is more adapted to testing rules in the lexer?

camilstaps commented on lib/rouge/lexers/clean.rb:
```diff
@@ -0,0 +1,130 @@
+      state :common_literals do
+        rule %r/'(?:[^'\\]|\\(?:x[0-9a-fA-F]+|\d+|.))'/, Str::Char
+
+        rule %r/[+~-]?\d+\.\d+(?:E[+-]?\d+)?\b/, Num::Float
```
Nice catch, but indeed, a leading digit and uppercase `E` are required. I have added examples.

camilstaps commented on lib/rouge/lexers/clean.rb:
```diff
@@ -0,0 +1,130 @@
+
+        rule %r/[+~-]?\d+\.\d+(?:E[+-]?\d+)?\b/, Num::Float
+        rule %r/[+~-]?\d+E[+-]?\d+\b/, Num::Float
+        rule %r/[+~-]?\d+/, Num::Integer
```
These are literals common to Clean and the intermediate language ABC. Because ABC is generated by the compiler and not meant to be written manually (except for short snippets like in `code { ... }`), it is more restrictive in what patterns it allows. I've added a comment about this to the top of this state.

camilstaps commented on lib/rouge/lexers/clean.rb:
```diff
@@ -0,0 +1,130 @@
+        mixin :common_literals
+        rule %r/[+~-]?0[0-7]+/, Num::Oct
+        rule %r/[+~-]?0x[0-9a-fA-F]+\b/, Num::Hex
+        rule %r/(\[)(\s*)(')(?=.*'\])/ do
```
No, the greediness was not intentional. It should behave the same, but the greedy version would be slower, correct? Otherwise I may be misunderstanding Ruby regexes. This has been changed, and an example has been added.

pyrmont replied:
Yeah, a lazy greedy match should (I think) be faster. Thanks for the change :)

camilstaps commented on lib/rouge/lexers/clean.rb:
```diff
@@ -0,0 +1,130 @@
+        mixin :common_literals
+      end
+
+      state :abc do
```
The `code { ... }` block allows including inline ABC code which is the intermediate language to which the compiler compiles. See https://en.wikipedia.org/wiki/Clean_(programming_language)#The_ABC-Machine for details; I also added a comment about this to the source code. The name derives from the three stacks that the abstract ABC machine has: for **a**rguments, **b**asic values and a **c**ontrol stack.

camilstaps commented on lib/rouge/lexers/clean.rb:
```diff
@@ -0,0 +1,130 @@
+
+      state :string do
+        rule %r/"/, Str::Double, :pop!
+        rule %r/\\./, Str::Double
```
I tried to cheat and not have to write out all the escape patterns :grin: They have been added now. It is correct that escaping `\x` and `\d` (for decimals) are case-insensitive while the integer pattern `0x` is not.

About unicode: I always get confused about encodings. Clean does not have this concept; the string `"°"` is legal but is really the array `{'\xc2','\xb0'}` (and `size "°"` is 2, not 1). I have added a test case to the visual sample for this and it seems to be fine, but maybe I'm missing something? There are, at least, no `\u` escapes as e.g. in JSON.

pyrmont replied:
Your examples look good to me!

camilstaps commented on lib/rouge/lexers/clean.rb:
```diff
@@ -0,0 +1,130 @@
+        mixin :abc_basic
+
+        rule %r/}/, Comment::Preproc, :pop!
+        rule %r/\.\w*/, Keyword, :abc_rest_of_line
```
I don't think `.` should be seen as punctuation. It is normally considered part of the keyword -- though you could argue indeed that `.` only tells you that a keyword is following. But because the keywords are usually short (`.d`, `.o`, `.n`, etc.), I think using separate classes would be distracting for the reader, what do you think?

pyrmont replied:
Yep, sounds cool to leave it as part of the keyword.

camilstaps commented:
Thanks @pyrmont. I believe I addressed all your questions and added examples (for the latter I did not reply to your comments). The sample has also been shortened significantly. I'm only unsure what to do with the `.` of ABC keywords.

camilstaps commented:
To be clear, Clean's ABC language is different from the musical notation language for which a lexer was proposed in https://github.com/rouge-ruby/rouge/pull/1092.

pyrmont commented:
@camilstaps Thanks for replying so quickly! Will try to get back to you with similar speed! :)

pyrmont commented on lib/rouge/lexers/clean.rb:
```diff
@@ -0,0 +1,130 @@
+      state :basic do
+        rule %r/\s+/m, Text::Whitespace
+
+        rule %r/\/\/\*.*/, Comment::Doc
```
The example in the document says that multiline doc comments begin with `/**` but there is no rule that matches that pattern. Instead, it's tokenised as a multiline comment. Is that intentional?

camilstaps replied:
No, that was an oversight. Thanks for catching it. The `:comment` and `:comment_doc` states now contain some duplicate code, is there a neat way around this?

pyrmont replied:
It's sometimes possible to avoid duplication by having rules be mixed in but that won't help here where you want to use different tokens. We have some lexers where people use some of Ruby's metaprogramming tricks to avoid duplicating code but I'm of the view that the obfuscation this results in outweighs any benefit from removing duplicate code .

In short, I don't think a little duplication is probelmatic, especially when it makes maintenance easier (which I think it does in this case).

pyrmont replied:
If the above is satisfactory from your side, the code looks ready to merge :)

camilstaps replied:
Yes, for me it's fine :) thanks for your time and the in-depth review!

pyrmont commented on lib/rouge/lexers/clean.rb:
```diff
@@ -0,0 +1,130 @@
+      state :root do
+        mixin :basic
+
+        rule %r/code(\s+inline)?\s*{/, Comment::Preproc, :abc
```
I don't think this has been added yet.

camilstaps replied:
Sorry, I must have removed it again when I shortened the sample. It should be there again now.

pyrmont commented:
@camilstaps I think there are only two comments outstanding at this stage. Let me know if anything is unclear!

pyrmont commented:
Thanks for your submission, @camilstaps! Always good to have another language be part of Rouge :)

We're currently following a two-week cadence for minor releases and this code is scheduled to land in v3.10.0 on Tuesday 3 September :tada:
